File: problems/334/334.increasing-triplet-subsequence.go
/*
 * @lc app=leetcode id=334 lang=golang
 *
 * [334] Increasing Triplet Subsequence
 */

package solution

// @lc code=start
import "math"

func increasingTriplet(nums []int) bool {
	// O(N) time, O(1) space

	first, second := math.MaxInt32, math.MaxInt32
	for _, third := range nums {
		if third > second {
			return true
		}

		if third > first {
			second = third
		} else {
			first = third
		}
	}

	return false
}

// @lc code=end

File: problems/334/334.increasing-triplet-subsequence.py
#
# @lc app=leetcode id=334 lang=python3
#
# [334] Increasing Triplet Subsequence
#


# @lc code=start
class Solution:
    def increasingTriplet(self, nums: List[int]) -> bool:
        # i < j < k and nums[i] < nums[j] < nums[k]
        # greedy

        first, second = float("inf"), float("inf")
        for third in nums:
            if third > second:
                return True

            if third > first:
                second = third
            else:
                first = third

        return False


# @lc code=end
